### DISCUSSION TITLE
Changing Dependency context for UI Testing

### DISCUSSION THREAD

**Discussion by YamiDaisuke** [pointfreeco/swift-composable-architecture#1583] [Q&A]
We have a project using the old Reducer and wanted to migrate to the ReducerProtocol. But we have a problem running our UI Tests. Previously we use this:

```swift
    static var environment: AppEnvironment {
        #if RELEASE
            return AppEnvironment.live
        #else
            if ProcessInfo.isUITesting || ProcessInfo.isUnitTesting {
                return AppEnvironment.test
            } else {
                return AppEnvironment.live
            }
        #endif
    }
```
Now we are trying to rely on the Dependency context which works perfect for Previews and Unit tests but for UI Testing we had to do this to make it work:

```swift
private lazy var store: StoreOf<AppReducer> = {
        if ProcessInfo.isUITesting {
            return StoreOf<AppReducer>(
                initialState: .loading,
                reducer: AppReducer()
                    .dependency(\.context, .test)
                    .dependency(\.date, DateGenerator { Date() })
                    .dependency(\.mainQueue, .immediate)
            )
        } else {
            return StoreOf<AppReducer>(
                initialState: .loading,
                reducer: AppReducer()
            )
        }
    }()
```

We are not sure if there is a better approach. We don't like that we have to provide alternatives to the dependencies provided by TCA (E.G.: `mainQueue`, `date`) to avoid falling into the unimplemented versions, so maybe there is a better way to do it. We use the `test` context for UI Test because we don't want to reach our backend we want to have fixed and predictable values

**Comment by mbrandonw** [comment]
Hi @YamiDaisuke, thanks for bringing this up. We have only used XCUITest a small amount, so we have some questions.

What are the problems you are having while running the UI test? Are you getting test failures from the dependencies? Or is it because the test value of `@Dependency(\.mainQueue)` doesn't actually execute its work?

**Reply by YamiDaisuke** [reply]
Hi, without using the explicit `.dependency(\.context, .test)` our test run but it uses live dependencies so it fails because our live API won't return the exact expected values. 

After overriding the context with `.test` we got this error: 
```
@Dependency(\.mainQueue) - An unimplemented scheduler scheduled an action to run immediately.
```
And the app does not execute the work. 

The error makes sense based on the idea that for unit testing this will catch when a reducer reaches out for an unexpected dependency but I think that for UI Test where we are most likely running the entire app I think there should be an easier way to make all the dependencies use the right test version.

**Reply by mbrandonw** [reply]
Interesting, since UI tests work by running two processes, one for the host app and one for the test, I'm surprised that `XCTFail`s in the _host_ process can cause failures in the test process.

Can you try pointing TCA to `main` and let us know how it differs, if at all? We made a few small changes to how test dependencies are accessed, and are curious if it changes anything for you.

Independent of that, I'm realizing that we should update our `unimplemented` scheduler to behave more like an immediately scheduler. Right now it just performs an `XCTFail` and then does nothing.

**Reply by mbrandonw** [reply]
We just cut a [release](https://github.com/pointfreeco/combine-schedulers/releases/tag/0.9.0) of combine-schedulers that make unimplemented schedulers behave like immediate schedulers, so that may help a bit too.

**Reply by YamiDaisuke** [reply]
> Interesting, since UI tests work by running two processes, one for the host app and one for the test, I'm surprised that `XCTFail`s in the _host_ process can cause failures in the test process.
> 
> Can you try pointing TCA to `main` and let us know how it differs, if at all? We made a few small changes to how test dependencies are accessed, and are curious if it changes anything for you.
> 
> Independent of that, I'm realizing that we should update our `unimplemented` scheduler to behave more like an immediately scheduler. Right now it just performs an `XCTFail` and then does nothing.

It's weird it nos actually breaking by the XCTFail but I caught that in the logs, the app is just not doing the work so the test fails. I'll try with main and let you know if that helps, but the change in combine-schedulers sounds like it will work.

**Reply by mbrandonw** [reply]
> It's weird it nos actually breaking by the XCTFail but I caught that in the logs, the app is just not doing the work so the test fails.

Ah ok, that's how I expected things to work. The `XCTFail`s are not causing the failures (you can ignore the logs), and instead what is causing problems is that the "unimplemented" main queue doesn't ever execute its work. I think if you update your application to use 0.9.0 of combine-schedulers things should work.

**Reply by YamiDaisuke** [reply]
ok! got it, it makes sense. I'll try the new version! thanks for the help!
